### ISSUE TITLE
H-357: Allow `change_context` to call `into_report`

### ISSUE THREAD

**Issue by bjchambers** [hashintel/hash#1968]
### Related Problem

I find myself often writing code that looks like:

```
some_call()
  .into_report()
  .change_context(Error::A)?;
some_other_call()
  .into_report()
  .change_context(Error::B)?;
...
```

This arises when calling libraries that don't return errors using error stack (or worse -- do so using `anyhow` since I'm in the middle of migrating).

### Proposed Solution

I think it should be possible to implement `change_context` for types that are not already `Report<E>`. For instance, if reports and `anyhow::Error` could be used with `IntoReport`, then the `ResultExt` trait could be implemented for any result containing an error that implements `IntoReport`. This would allow the original code to be written as:

```
some_call()
  .change_context(Error::A)?;
some_other_call()
  .change_context(Error::B)?;
...

### Alternatives

_No response_

### Additional context

_No response_

**Comment by TimDiekmann** [comment]
The situation is a bit tricky. `into_report()` or `Report::new()` does more than just put an error `E` into `Report<E>`. If necessary, a `Backtrace` or even a `SpanTrace` will be captured. `Report::change_context` (or the corresponding method on the `Result`) will “only” capture a new `Frame` and put it in the report, which should be pretty cheap. So calling `change_context()` directly on a non-`Report`-`Result` would hide from the user that possibly more happens than putting the error into a new frame. We have assumed `into_report` so far because it is more explicit to predict the behavior. In fact, `into_report()` should not be called at all when in the hot path, so this precaution is perhaps a bit too verbose.

What I personally would like to see is a method that does both, i.e. an `into_report_with_context<C>(C)`, but I have not been able to find a name for this method that is sufficiently short (obviously, the above is too long for a method that is needed relatively often), but still describes the behavior well enough, since this method is defined on `Result<T, E>` and that is an enormously frequently used type. Also, the interface is getting even bigger.

Personally, I don't have a strong opinion on this and I know that `into_report().change_context(C)` is quite annoying. However, from personal experience, I have found that the migration itself is quite difficult as many types rely on non-`error-stack` types so you have to call `into_report()` quite often, but this is better once you can rely on your own functions that rely on `error-stack` (and of course any other crate, which using `error-stack`) and I quite like the compile-time checked validation, that the error is wrapped in a `Report`.

Happy to hear your thoughts on that!

---

Btw, just in case you didn't know: `error-stack` provides [`IntoReportCompat`](https://docs.rs/error-stack/latest/error_stack/trait.IntoReportCompat.html) to allow `anyhow::Error` or `eyre::Report` to be used directly. This does not skip the `into_report()` step, but it lets you treat the type as it would implement `Error` or `Context`.

**Comment by mzruya** [comment]
I've been experimenting with error-stack over the weekend, converting a small project from eyre. I stumbled across this issue as I also find the boilerplate to be distracting.

The project I'm converting has a series of `std::io`/`serde` calls, simple statements such as:
```rust
let config = serde_yaml::from_reader(File::open(&config_path)?)?
```

Now become:
```rust
let config_file = File::open(&config_path).into_report().change_context(Error::Io)?;
let config = serde_yaml::from_reader(config_file).into_report().change_context(Error::SerdeYml)?
```

Not the worst, but does feel redundant and hurts readability. I made a small extension trait that combines them into a single call. `into_report` => `change_context` => `into_context`:

```rust
use error_stack::{Context, Report, Result, IntoReport, ResultExt};

/// Extension trait to shorten repretitive calls, `into_report().change_context(NewError) => `into_context(NewError)`
pub trait IntoContext: Sized {
    /// Type of the [`Ok`] value in the [`Result`]
    type Ok;

    /// Type of the resulting [`Err`] variant wrapped inside a [`Report<E>`].
    type Err;

    /// Converts the [`Err`] variant of the [`Result`] to a [`Report<C>`]
    fn into_context<C>(self, context: C) -> Result<Self::Ok, C>
    where
        C: Context;
}

impl<T, E> IntoContext for core::result::Result<T, E>
where
    Report<E>: From<E>,
{
    type Err = E;
    type Ok = T;

    fn into_context<C>(self, context: C) -> Result<T, C>
    where
        C: Context,
    {
        self.into_report().change_context(context)
    }
}

```

Which simplifies the conversion to:
```rust
let config_file = File::open(&config_path).into_context(Error::Io)?;
let config = serde_yaml::from_reader(config_file).into_context(Error::SerdeYml)?
```

**Comment by mzruya** [comment]
I've been experimenting with error-stack by converting a small project from eyre. I stumbled across this issue as I also find the boilerplate to be distracting.

The project I'm converting has a series of `std::io`/`serde` calls, simple statements such as:
```rust
let config = serde_yaml::from_reader(File::open(&config_path)?)?
```

Now become:
```rust
let config_file = File::open(&config_path).into_report().change_context(Error::Io)?;
let config = serde_yaml::from_reader(config_file).into_report().change_context(Error::SerdeYml)?
```

Not the worst, but does feel redundant and hurts readability. I made a small extension trait that combines them into a single call. `into_report` => `change_context` => `into_context`:

```rust
use error_stack::{Context, Report, Result, IntoReport, ResultExt};

/// Extension trait to shorten repetitive calls, `into_report().change_context(NewError) => `into_context(NewError)`
pub trait IntoContext: Sized {
    /// Type of the [`Ok`] value in the [`Result`]
    type Ok;

    /// Type of the resulting [`Err`] variant wrapped inside a [`Report<E>`].
    type Err;

    /// Converts the [`Err`] variant of the [`Result`] to a [`Report<C>`]
    fn into_context<C>(self, context: C) -> Result<Self::Ok, C>
    where
        C: Context;
}

impl<T, E> IntoContext for core::result::Result<T, E>
where
    Report<E>: From<E>,
{
    type Err = E;
    type Ok = T;

    fn into_context<C>(self, context: C) -> Result<T, C>
    where
        C: Context,
    {
        self.into_report().change_context(context)
    }
}

```

Which simplifies the conversion to:
```rust
let config_file = File::open(&config_path).into_context(Error::Io)?;
let config = serde_yaml::from_reader(config_file).into_context(Error::SerdeYml)?
```

**Comment by TimDiekmann** [comment]
Hi @mzruya and sorry for responding so late.

I'm not 100% sure the name is well suited for such a trait you don't convert the context into another context but you still create a report and put the context into it. 

We agreed on the change proposed in the OP to allow `change_context()` to implicitly call `into_report()`. This is probably the most asked question/request for `error-stack` and as this is not expected to happen in the hot path this will hopefully improve the DX. We consider removing the `IntoReport` trait completely as a result as the `ResultExt` trait will be implemented on the same types. This will make the overall `error-stack` interface easier (this would be done via a deprecation in the next release and removal in the release afterward if it's possible to remove).

**Comment by TimDiekmann** [comment]
The issue is now available in [`error-stack@0.4.0`](https://github.com/hashintel/hash/releases/tag/error-stack%400.4.0)
